airflow/models/dagbag.py

```diff
@@ -702,7 +702,7 @@ def needs_perms(dag_id: str) -> bool:
                     return True
             return False
 
-        if dag.access_control or needs_perms(root_dag_id):
+        if dag.access_control or not needs_perms(root_dag_id):
```

Comment:
I don't think this logic is correct. You are saying "or dag perms already exist", a.k.a sync perms for existing DAGs, but don't we need to do it for new DAGs? You probably want to just remove this if block completely so we sync perms in both cases. This is effectively reverting #15464.

I wonder if we can figure out a better way to check that isn't as costly? I'm okay with extra cost for access_control, but ideally we don't impact DAGs that don't use it.

Either way, a good first step is to write a test case that fails with the current code in main. We will want it anyways before we merge this.

Reply:
Hi @jedcunningham,
I just removed `if dag.access_control or needs_perms(root_dag_id):` and `if access_control:`. This is effectively reverting https://github.com/apache/airflow/pull/15464. But https://github.com/apache/airflow/issues/25149 will be resolved, I think.